src/lib/io/sol-iio.c

```diff
@@ -595,22 +615,25 @@ channel_get_pure_name(struct sol_iio_channel *channel)
             channel_pure_name = calloc(1, channel_name_len + 1);
             original_channel_pure_name = channel_pure_name;
             for (i = 0; i < channel_name_len; i++) {
-                if (isalpha((uint8_t)channel->name[i]) || channel->name[i] == '-' || channel->name[i] == '_')
-                    *channel_pure_name++ = channel->name[i];
-                else if (isdigit((uint8_t)channel->name[i])) {
+                if (isalpha((uint8_t)channel_name[i]) || channel_name[i] == '-'
+                    || channel_name[i] == '_')
+                    *channel_pure_name++ = channel_name[i];
+                else if (isdigit((uint8_t)channel_name[i])) {
                     modified = true;
                     continue;
                 }
             }
-            if (modified)
+            if (modified) {
+                free(internsity_channel_name);
                 return original_channel_pure_name;
+            }
 
             free(original_channel_pure_name);
-            return NULL;
+            return internsity_channel_name;
         }
     }
 
-    return NULL;
+    return internsity_channel_name;
```

Comment:
I prefer to keep the current complicated implement, in order to resolve the issue #2244. I will refactor this function in the next PR